app/src/main/java/ch/hikemate/app/model/elevation/ElevationServiceRepository.kt

```diff
@@ -129,10 +287,60 @@ class ElevationServiceRepository(private val client: OkHttpClient) : ElevationSe
         val elevationResponse = decodeFromString<ElevationResponse>(body)
 
         // Call the onSuccess callback with the list of elevations
-        onSuccess(elevationResponse.results.map { it.elevation })
+        onSuccess(elevationResponse.elevations)
+        failedRequests = 0
       } else {
-
-        onFailure(Exception("Failed to get elevation"))
+        when (response.code) {
+          500,
+          504,
+          429 -> {
+            failedRequests++
+            Log.e(
+                "ElevationServiceRepository",
+                "Failed to get elevation. Status code: ${response.code}, retrying, failedRequests: $failedRequests")
+            if (failedRequests > MAX_FAILED_REQUESTS) {
+              onFailure(Exception("Failed to get elevation. Status code: ${response.code}"))
+              failedRequests = 0
+              return
+            } else {
+              CoroutineScope(dispatcher).launch {
+                delay(failedRequests * FAILED_REQUEST_DELAY)
+                launchUpdateOfCache()
+              }
+            }
+          }
+          413 -> {
+            Log.e(
+                "ElevationServiceRepository",
+                "Failed to get elevation. Status code: ${response.code}, too many coordinates")
+            CoroutineScope(dispatcher).launch {
+              var firstHalf: List<ElevationRequest>
+              var secondHalf: List<ElevationRequest> = emptyList()
+              Log.d("ElevationServiceRepository", "Split requests in half")
+              mutex.withLock {
+                val halves = requests.partition { requests.indexOf(it) <= requests.size / 2 }
+                firstHalf = halves.first
+                secondHalf = halves.second
+                requests.clear()
+                requests.addAll(firstHalf)
+                Log.d("ElevationServiceRepository", "First half size: ${firstHalf.size}")
+              }
+              launchUpdateOfCache(
+                  onSuccess = {
+                    CoroutineScope(dispatcher).launch {
+                      mutex.withLock {
+                        Log.d("ElevationServiceRepository", "Second half size: ${secondHalf.size}")
+                        requests.addAll(secondHalf)
+                      }
+                      launchUpdateOfCache()
+                    }
+                  })
```

Comment:
`requests` is accessed only under the mutex lock, thus, it can't be modified when splitting. However, one can add requests in between the `firstHalf` and the `secondHalf`, and thus, the requests will be processed with the `secondHalf`. If there is again too many requests, the process will repeat and resplit